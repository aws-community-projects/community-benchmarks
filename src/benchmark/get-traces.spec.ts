import { BatchGetTracesCommand, XRayClient } from '@aws-sdk/client-xray';
import { mockClient } from 'aws-sdk-client-mock';

import trace from '../../sample-trace.json';
import { handler } from './get-traces';

const xrayMock = mockClient(XRayClient);

const mockFnResult = {
  SdkHttpMetadata: {
    HttpHeaders: {
      ['X-Amzn-Trace-Id']:
        'root=FAKE_TRACE_ID;parent=1234567890abcdef;sampled=1',
    },
  },
};

beforeEach(() => xrayMock.reset());

describe('get-traces function', () => {
  test('fetch and parse traces', async () => {
    xrayMock.on(BatchGetTracesCommand).resolves(trace);

    const result = await handler({
      MapResult: Array(10).fill(mockFnResult),
      token: 'FAKE_TASK_TOKEN',
    });
    expect(result).toMatchSnapshot();
  });

  test('throw exception if no traces', async () => {
    xrayMock.on(BatchGetTracesCommand).resolves({ Traces: [] });
    try {
      await handler({
        MapResult: Array(10).fill(mockFnResult),
        token: 'FAKE_TASK_TOKEN',
      });
    } catch (e) {
      expect((e as Error).message).toBe('No traces!');
    }
  });

  test('throw exception if not all traces have arrived', async () => {
    xrayMock.on(BatchGetTracesCommand).resolves({ Traces: [{}] });
    try {
      await handler({
        MapResult: Array(10).fill(mockFnResult),
        token: 'FAKE_TASK_TOKEN',
      });
    } catch (e) {
      expect((e as Error).message).toBe('In Progress');
    }
  });
});
