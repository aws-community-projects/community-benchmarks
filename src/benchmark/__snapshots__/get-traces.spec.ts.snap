// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-traces function fetch and parse traces 1`] = `
Array [
  Object {
    "averageColdStart": 504.18,
    "averageDuration": 2413.51,
    "coldStartPercent": 50,
    "durations": Array [
      2362.4744415283203,
      2464.5466804504395,
    ],
    "inits": Array [
      504.1818618774414,
      0,
    ],
    "name": "csv2ddb-sdk2-layer",
    "p90ColdStart": 504.18,
    "p90Duration": 2454.34,
  },
  Object {
    "averageColdStart": 494.13,
    "averageDuration": 2475.8,
    "coldStartPercent": 50,
    "durations": Array [
      2351.0706424713135,
      2600.536823272705,
    ],
    "inits": Array [
      0,
      494.13204193115234,
    ],
    "name": "csv2ddb-sdk2",
    "p90ColdStart": 494.13,
    "p90Duration": 2575.59,
  },
  Object {
    "averageColdStart": 512.29,
    "averageDuration": 2723.27,
    "coldStartPercent": 50,
    "durations": Array [
      2546.1673736572266,
      2900.376796722412,
    ],
    "inits": Array [
      0,
      512.2928619384766,
    ],
    "name": "csv2ddb-sdk3",
    "p90ColdStart": 512.29,
    "p90Duration": 2864.96,
  },
  Object {
    "averageColdStart": 492.53,
    "averageDuration": 2657.59,
    "coldStartPercent": 50,
    "durations": Array [
      2432.806968688965,
      2882.368803024292,
    ],
    "inits": Array [
      0,
      492.52843856811523,
    ],
    "name": "csv2ddb-sdk2-native",
    "p90ColdStart": 492.53,
    "p90Duration": 2837.41,
  },
  Object {
    "averageColdStart": 490.24,
    "averageDuration": 2565.17,
    "coldStartPercent": 50,
    "durations": Array [
      2445.6546306610107,
      2684.6818923950195,
    ],
    "inits": Array [
      0,
      490.236759185791,
    ],
    "name": "csv2ddb-sdk2-clients",
    "p90ColdStart": 490.24,
    "p90Duration": 2660.78,
  },
]
`;
